### DISCUSSION TITLE
How to handle dependencies between search params, persistence, and share logic between routes?

### DISCUSSION THREAD

**Discussion by szimek** [TanStack/router#541] [Q&A]
Hi!

I'm working on an app that I think is somewhat similar-ish to nozzle.io, i.e. we've got a lot of reports and dashboards that are configured via search params. That's why it's really exciting that there's finally a router library that focuses on proper support for search params!

We have some maybe less common requirements, and I was wondering how might one implement them when using TanStack router:

1. Handling dependencies between search params
In our web app we have cases where one search param might depend on another. E.g. we have a page that shows data about some mobile apps and users can switch between app stores (Google Play or App Store) and app categories. Available categories depend on the selected app store, and when the user changes the store, we try to map the selected category to its respective category in the other store. If we can't, we set it to the default value for the selected store.
An additional requirement is that we want to update all dependent search params (in this case, the ones for store and category) in one call, so that the browser's back button still works correctly and doesn't navigate the user to some intermediate state with invalid params.

2. Returning a reason when validation fails
In some cases, when a search param has a non-empty value that fails the validation, we want to set it to a valid value, still render the requested page rather than an error component, and also notify the user that this param was modified by us. An example might be when the user navigates to a page that shows the details of an app in a country, but the selected app is not available in the selected country - in this case we might want to switch to the default country, and notify the user about it. Is that something that is possible with the current `validateSearch` API?

4. Persistence of search params
Another use case that we need to support is that when a user navigates between pages, we want to store the search params somewhere, so that we can reuse them on other  pages. One scenario might be e.g. the following:
    - user navigates to a report page for multiple apps and selects a few apps - their ids are stored in search params
    - user navigates to a details page for one of the selected apps - only the id of that one app is present in search params
    - user navigates to another report for multiple apps - in this case we'd like to show the same apps they had selected in the first step

    In general, the ideal solution would be if we could automatically update all links to have the correct search string when user navigates between pages, but given that it might be not feasible/possible in many cases, it's also ok if the search params are resolved only after the user clicks a link.

5. Shared logic
Finally, given that for some search params the default value, validation and (de)serialization logic would be exactly the same on many routes/pages, it would be great if there was an easy way to share it.

It's a really interesting and fun challenge and there's even a bit more to it than this, e.g. when to use `replaceState` vs `pushState`, etc. At the moment, we're using React Router and we have a basic prototype that implements some of the above requirements, but it's far from perfect - e.g. the whole setup happens in the top-level page component, rather than at the route level, so we don't have correct types for search params in links. We also use zustand to persist search params across pages.

Cheers!

**Comment by tannerlinsley** [comment]
Let’s schedule a call to chat about this.

**Reply by szimek** [reply]
Hi @tannerlinsley ! Please let me know what's the best way to contact you to schedule this call. Thanks!

**Comment by szimek** [comment]
Hi @tannerlinsley, thanks for your reply! I reached out to you on Twitter - I have the same username there, i.e. @szimek.

**Comment by tordans** [comment]
I am very interested in learning more about those topics as well. And – especially after reading https://twitter.com/tannerlinsley/status/1600390105755439104 – I would like to add to the list…

- How to handle url param migrations / version
- How to keep the URL short when complex objects are used to handle the state (ATM I am using [JSURL2](https://github.com/TanStack/router/blob/beta/docs/guide/custom-search-param-serialization.md#using-jsurl2) coupled with some custom code that shortens object keys based on a custom table

Maybe those can be topics for a (corporate) blogpost, livestream or something like this?

**Comment by visualjerk** [comment]
As an addition to 4., it would be great to have a way of using `useSearch` in a component, that is used in multiple routes without relying on `{ strict: false }`.

Example of a possible API:

```ts
const searchParams = useSearch({
  from: ['/route-a', '/route-b']
})
```

**Comment by Rendez** [comment]
Hi there. Where there any learnings/conclusions on the original questions?
